src/types/walk.rs

```diff
@@ -5,63 +5,149 @@ extern crate fraction;
 extern crate num_traits;
 extern crate petgraph;
 
-#[derive(Debug, Default, PartialEq, Eq)]
-pub struct RandomWalks<Id> {
-    random_walks_internal: Vec<RandomWalk<Id>>,
+use std::collections::HashMap;
+use std::hash::Hash;
+
+#[derive(Debug, Default)]
+pub struct RandomWalks<Id>
+where
+    Id: Hash + Eq,
+{
+    /// A `HashMap` between the source of the walk and the walk itself.
+    random_walks: HashMap<Id, RandomWalk<Id>>,
 }
 
 impl<Id> RandomWalks<Id>
 where
-    Id: PartialEq,
+    Id: Clone + Eq + Hash,
 {
     pub fn new() -> Self {
         RandomWalks {
-            random_walks_internal: Vec::new(),
+            random_walks: HashMap::new(),
         }
     }
 
     pub fn add_walk(&mut self, walk: RandomWalk<Id>) {
-        self.random_walks_internal.push(walk);
+        self.random_walks
+            .insert(walk.random_walk_source.clone(), walk);
```

Comment:
Doesn't this overwrite existing walks starting in `random_walk_source`? https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.insert

Reply:
Good catch, it does, but I have later realised using a `HashMap` wasn't buying us anything, so in a more recent revision of the code I have reverted it back to be a `Vec<RandomWalk>`. Unfortunately, no changes in the timing of `cargo bench` (if not a minor speedup) 😢